pkg/urbit/jets/e/aes_siv.c

```diff
@@ -0,0 +1,357 @@
+    u3r_bytes_reverse(0, ad_w, ad_y, u3h(ads));
+
+    c3_w ret = AES_SIV_AssociateData(ctx, ad_y, ad_w);
+    u3a_free(ad_y);
+
+    if ( 0 == ret ) {
+      AES_SIV_CTX_free(ctx);
+      return u3_none;
+    }
+
+    ads = u3t(ads);
+  }
+
+  c3_w txt_w = u3r_met(3, txt);
+  c3_y* txt_y = u3a_malloc(txt_w);
+  u3r_bytes_reverse(0, txt_w, txt_y, txt);
+
+  c3_w iv_w = 16;
+  c3_y* iv_y = u3a_malloc(iv_w);
+  c3_w out_w = txt_w;
+  c3_y* out_y = u3a_malloc(out_w);
+  if ( 0 == AES_SIV_EncryptFinal(ctx, iv_y, out_y, txt_y, txt_w) ) {
+    u3a_free(iv_y);
+    u3a_free(out_y);
+    u3a_free(txt_y);
+    AES_SIV_CTX_free(ctx);
+    return u3_none;
+  }
+
+  u3a_free(txt_y);
+  AES_SIV_CTX_free(ctx);
+
+  // Read the first 16 bytes as the "iv"
+  u3_noun iv_n = u3i_bytes(16, iv_y);
+  u3_noun msg_n = u3i_bytes(txt_w, out_y);
+
+  // Reverse byte order for output
+  u3_noun rev_iv_n = u3qc_rev(3, iv_w, iv_n);
+  u3_noun rev_msg_n = u3qc_rev(3, txt_w, msg_n);
+
+  u3a_free(iv_y);
+  u3a_free(out_y);
+  u3z(iv_n);
+  u3z(msg_n);
+
+  return u3nt(rev_iv_n, txt_w, rev_msg_n);
+}
+
+static u3_noun _siv_de(c3_y* key_y,
+                       c3_w keysize,
+                       u3_atom ads,
+                       u3_atom iv,
+                       u3_atom len,
+                       u3_atom txt)
+{
+  AES_SIV_CTX* ctx = AES_SIV_CTX_new();
+  if ( 0 == ctx ) {
+    return u3_none;
+  }
+
+  if ( 0 == AES_SIV_Init(ctx, key_y, keysize) ) {
+    AES_SIV_CTX_free(ctx);
+    return u3_none;
+  }
+
+  while (u3_nul != ads) {
+    c3_w ad_w = u3r_met(3, u3h(ads));
+    c3_y* ad_y = u3a_malloc(ad_w);
+    u3r_bytes_reverse(0, ad_w, ad_y, u3h(ads));
+
+    c3_w ret = AES_SIV_AssociateData(ctx, ad_y, ad_w);
+    u3a_free(ad_y);
+
+    if ( 0 == ret ) {
+      AES_SIV_CTX_free(ctx);
+      return u3_none;
+    }
+
+    ads = u3t(ads);
+  }
+
+  c3_w txt_w = u3r_word(0, len);  // ?
+  c3_y* txt_y = u3a_malloc(txt_w);
+  u3r_bytes_reverse(0, txt_w, txt_y, txt);
+
+  c3_w iv_w = 16;
+  c3_y* iv_y = u3a_malloc(iv_w);
+  u3r_bytes_reverse(0, 16, iv_y, iv);
+
+  c3_y* out_y = u3a_malloc(txt_w);
+  if ( 0 == AES_SIV_DecryptFinal(ctx, out_y, iv_y, txt_y, txt_w) ) {
+    u3a_free(iv_y);
+    u3a_free(out_y);
+    u3a_free(txt_y);
+    AES_SIV_CTX_free(ctx);
+
+    // Either decryption failed or signature bad or there was a memory
+    // error. Some of these are deterministic and some are not. return u3_none
+    // to fallback to the Nock implementation.
+    return u3_none;
+  }
+
+  u3a_free(txt_y);
+  AES_SIV_CTX_free(ctx);
+
+  // Read the first 16 bytes as the "iv"
+  u3_noun msg_n = u3i_bytes(txt_w, out_y);
+
+  // Reverse byte order for output
+  u3_noun rev_msg_n = u3qc_rev(3, txt_w, msg_n);
+
+  u3a_free(iv_y);
+  u3a_free(out_y);
+  u3z(msg_n);
+
+  return u3nc(0, rev_msg_n);
+}
+
+
+u3_noun
+u3qea_siva_en(u3_atom key,
+              u3_atom ads,
+              u3_atom txt)
+{
+  c3_y key_y[32];
+  AES_KEY key_u;
+  c3_assert(u3r_met(3, key) <= 32);
```

Comment:
This is a `bail:oops`, which could theoretically be handled like any `bail:fail`, but currently aborts the process. What is the hoon/nock behavior in this case?

Reply:
Same for all subsequent `c3_assert()` calls.

Reply:
This pattern was taken from the other AES jets. There actually are no sizing checks anywhere in the hoon code. I've changed these to also be partial failures and to fallback to the nock.